python/graphstorm/inference/np_infer.py

```diff
@@ -168,3 +171,90 @@ def infer(self, loader, save_embed_path, save_prediction_path=None,
+        inference on Amazon SageMaker endpoints.
+
+    The real-time inferrer has three major differences from the node prediction offline inferrer:
+
+    1. directly setting and using a DGL graph as input to extract features.
+    2. using DGL dataloader, instead of GS dataloaders that rely on DGL distributed graphs.
+    3. no evaluation section as there is no label on real-time inference.
+    """
+    def infer(self,
+              g,
+              dataloader,
+              infer_ntypes,
+              nfeat_fields,
+              efeat_fields=None,
+              return_proba=True):
+        """
+        ``GSGnnNodePredictionRealtimeInferrer`` defines the ``infer()`` method that performs
+        three tasks:
+
+        1. Extract one batch using the given dataloader;
+        2. Prepare input node and edge features;
+        3. Compute inference results for nodes with target node type and return the results.
+
+        Parameters
+        ----------
+        g: DGLGraph
+            The inference graph data in the format of a DGL heterograph. For built-in inference
+            pipeline, this graph should be constructed by using methods in the
+            `gconstruct.construct_payload_graph.py` file.
+        dataloader: DGL dataloader
+            A DGL dataloader class for node prediction, e.g., `DataLoader`.
+        infer_ntypes: list of string or a string
+            The list of the target node types. Or a single string of the target node type.
+        nfeat_fields: dict of {str: list}
+            The node feature fields in the format of a dict, whose keys are the node type names, and
+            values are lists of feature names.
+        efeat_fields: dict of {tuple: list}
+            The edge feature fields in the format of a dict, whose keys are the edge type name
+             tuples, and values are lists of feature names. Default is None.
+        return_proba: boolean
+            If return probability of model predictions. Default is True.
+
+        Returns
+        -------
+        predictions: dict
+            The inference results in the format of {str: tensor}.
+        """
+        assert isinstance(g, dgl.DGLGraph), 'The input graph of ' + \
+            '\"GSGnnNodePredictionRealtimeInferrer\" must be an instance of dgl.DGLGraph, ' + \
+            f'but got {type(g)}.'
+        assert isinstance(infer_ntypes, (list, str)), 'The value of \"infer_ntypes\" ' + \
+            f'should be either a list of strings or a single string, but got {infer_ntypes}.'
+        if isinstance(infer_ntypes, str):
+            infer_ntypes = [infer_ntypes]
+
+        # set model to be in the evaluation mode
+        self._model.eval()
+        # extract one mini-batch blocks using the given DGL dataloader
+        all_nodes = []
+        all_blocks = []
+        for input_nodes, _, blocks in dataloader:
+            all_nodes = input_nodes
```

Comment:
all_nodes.append?

You only store the last element from dataloader.

Reply:
Yes if we know the iterator only has one element we need to make that clear in the code. Maybe add an assertion as well

Reply:
Yes. revised to use `.next()`